lib/rspec/rails/matchers/have_reported_error.rb

```diff
@@ -0,0 +1,195 @@
+require "rspec/rails/matchers/base_matcher"
+
+module RSpec
+  module Rails
+    module Matchers
+      # @private
+      class ErrorSubscriber
+        attr_reader :events
+
+        ErrorEvent = Struct.new(:error, :attributes)
+
+        def initialize
+          @events = []
+        end
+
+        def report(error, **attrs)
+          @events << ErrorEvent.new(error, attrs.with_indifferent_access)
+        end
+      end
+
+      # Matcher class for `have_reported_error`. Should not be instantiated directly.
+      #
+      # @private
+      # @see RSpec::Rails::Matchers#have_reported_error
+      class HaveReportedError < RSpec::Rails::Matchers::BaseMatcher
+        def initialize(expected_error = nil)
```

Comment:
We usually make a distinction between an explicitly passed nil (sometimes by accident), and no argument. You can see the UNSPECIFIED const in the BaseMatcher for an example.